File: modulo-govsocial/api/app/schemas/beneficios.py
import uuid
from datetime import datetime
from decimal import Decimal
from typing import Optional

from pydantic import BaseModel, Field


# ── Estoque ───────────────────────────────────────────────────────────
class EstoqueCreate(BaseModel):
    unit_id: uuid.UUID
    benefit_type_code: str = Field(..., max_length=40)
    quantidade_inicial: Decimal = Field(..., ge=0, max_digits=12, decimal_places=2)
    quantidade_minima: Decimal = Field(Decimal("0"), ge=0, max_digits=12, decimal_places=2)
    unidade_medida: str = "UNIDADE"
    valor_unitario_referencia: Optional[Decimal] = None


class EstoqueUpdate(BaseModel):
    quantidade_atual: Optional[Decimal] = None
    quantidade_minima: Optional[Decimal] = None
    valor_unitario_referencia: Optional[Decimal] = None


class EstoqueOut(BaseModel):
    id: uuid.UUID
    unit_id: uuid.UUID
    benefit_type_code: str
    quantidade_atual: Decimal
    quantidade_minima: Decimal
    unidade_medida: str
    valor_unitario_referencia: Optional[Decimal]
    created_at: datetime
    updated_at: datetime


class EstoqueMovement(BaseModel):
    """Entrada ou saída manual de estoque (ajuste)."""
    quantidade: Decimal = Field(..., max_digits=12, decimal_places=2)
    observacao: Optional[str] = None


# ── Concessão de Benefício ────────────────────────────────────────────
class ConcessaoCreate(BaseModel):
    family_id: uuid.UUID
    person_id: Optional[uuid.UUID] = None
    unit_id: uuid.UUID
    benefit_type_code: str = Field(..., max_length=40)
    quantidade: Decimal = Field(Decimal("1"), ge=0, max_digits=12, decimal_places=2)
    valor_total: Optional[Decimal] = None
    solicitado_por_id: Optional[uuid.UUID] = None


class ConcessaoUpdate(BaseModel):
    """Atualização de campos antes da análise/aprovação."""
    quantidade: Optional[Decimal] = None
    valor_total: Optional[Decimal] = None
    person_id: Optional[uuid.UUID] = None


class ParecerCreate(BaseModel):
    """Emissão de parecer técnico."""
    parecer: Optional[str] = None


class AprovacaoCreate(BaseModel):
    """Aprovação da concessão."""


class NegacaoCreate(BaseModel):
    """Negação da concessão."""
    motivo_negacao: str


class EntregaCreate(BaseModel):
    """Registro de entrega com assinatura."""


class ConcessaoOut(BaseModel):
    id: uuid.UUID
    family_id: uuid.UUID
    person_id: Optional[uuid.UUID]
    unit_id: uuid.UUID
    benefit_type_code: str
    quantidade: Decimal
    valor_total: Optional[Decimal]
    status: str
    data_solicitacao: datetime
    data_analise: Optional[datetime]
    data_aprovacao: Optional[datetime]
    data_entrega: Optional[datetime]
    solicitado_por_id: Optional[uuid.UUID]
    analisado_por_id: Optional[uuid.UUID]
    aprovado_por_id: Optional[uuid.UUID]
    parecer: Optional[str] = None
    parecer_restrito: bool = False
    motivo_negacao: Optional[str]
    comprovante_gerado: bool
    assinatura_data: Optional[datetime]
    created_at: datetime
    updated_at: datetime


class ConcessaoListItem(BaseModel):
    id: uuid.UUID
    family_id: uuid.UUID
    unit_id: uuid.UUID
    benefit_type_code: str
    status: str
    data_solicitacao: datetime
    valor_total: Optional[Decimal]

File: modulo-govsocial/web/src/tipos/beneficios.ts
/** DTOs de benefícios eventuais — espelham app/schemas/beneficios.py. */

export type StatusConcessao =
  | "SOLICITADO"
  | "EM_ANALISE"
  | "APROVADO"
  | "ENTREGUE"
  | "NEGADO"
  | "CANCELADO";

export type ConcessaoOut = {
  id: string;
  family_id: string;
  person_id: string | null;
  unit_id: string;
  benefit_type_code: string;
  quantidade: number;
  valor_total: number | null;
  status: StatusConcessao;
  data_solicitacao: string;
  data_analise: string | null;
  data_aprovacao: string | null;
  data_entrega: string | null;
  solicitado_por_id: string | null;
  analisado_por_id: string | null;
  aprovado_por_id: string | null;
  // Parecer é sensível: só vem quando concedido (parecer_restrito sinaliza omissão).
  parecer: string | null;
  parecer_restrito: boolean;
  motivo_negacao: string | null;
  comprovante_gerado: boolean;
  assinatura_data: string | null;
  created_at: string;
  updated_at: string;
};

export type ConcessaoListItem = {
  id: string;
  family_id: string;
  unit_id: string;
  benefit_type_code: string;
  status: StatusConcessao;
  data_solicitacao: string;
  valor_total: number | null;
};

export type ConcessaoCreate = {
  family_id: string;
  person_id?: string | null;
  unit_id: string;
  benefit_type_code: string;
  quantidade: number;
  valor_total?: number | null;
  solicitado_por_id?: string | null;
};
